pkg/nsdns/manager.go

```diff
@@ -0,0 +1,169 @@
+package nsdns
+
+import (
+	"fmt"
+	"os"
+	"sync"
+)
+
+import (
+	log "github.com/sirupsen/logrus"
+	apinetworkingv1 "k8s.io/api/networking/v1"
+)
+
+import (
+	"github.com/Eagerod/kube-namesilo-dns/pkg/icanhazip"
+	"github.com/Eagerod/kube-namesilo-dns/pkg/namesilo_api"
+)
+
+type dnsManagerCache struct {
+	CurrentRecords   []namesilo_api.ResourceRecord
+	CurrentIpAddress string
+}
+
+func NewDnsManagerCache() *dnsManagerCache {
+	return &dnsManagerCache{
+		[]namesilo_api.ResourceRecord{},
+		"",
+	}
+}
+
+type DnsManager struct {
+	BareDomainName     string
+	TargetIngressClass string
+
+	Api namesilo_api.NamesiloApi
+
+	cacheLock              *sync.Mutex
+	cache                  *dnsManagerCache
+	RefreshesCacheOnUpdate bool
+}
+
+func NewDnsManager(domainName, ingressClass string) (*DnsManager, error) {
+	nsApiKey := os.Getenv("NAMESILO_API_KEY")
+	if nsApiKey == "" {
+		return nil, fmt.Errorf("failed to find NAMESILO_API_KEY in environment; cannot proceed")
+	}
+
+	return NewDnsManagerWithApiKey(domainName, ingressClass, nsApiKey)
+}
+
+func NewDnsManagerWithApiKey(domainName, ingressClass, apiKey string) (*DnsManager, error) {
+	if domainName == "" {
+		return nil, fmt.Errorf("must provide a domain name to target DNS record updates")
+	}
+
+	if ingressClass == "" {
+		return nil, fmt.Errorf("must provide an ingress class to generate DNS records")
+	}
+
+	api := namesilo_api.NewNamesiloApi(domainName, apiKey)
+
+	dm := DnsManager{
+		domainName,
+		ingressClass,
+		api,
+		&sync.Mutex{},
+		NewDnsManagerCache(),
+		false,
+	}
+
+	return &dm, nil
+}
+
+func (dm *DnsManager) ShouldProcessIngress(ingress *apinetworkingv1.Ingress) bool {
+	ic, ok := ingress.Annotations["kubernetes.io/ingress.class"]
+	if !ok {
+		return false
+	}
+
+	return ic == dm.TargetIngressClass
+}
+
+func (dm *DnsManager) HandleIngressExists(ingress *apinetworkingv1.Ingress) error {
```

Comment:
Don't really like the naming discrepancy here. 

```
HandleIngressExists
HandleIngressDeleted
```

Can't think of a super great alternative though, especially given that records that exist at the beginning of the run will all pass through the `Exists` flow just to ensure that the state is correct once the app has finished its first cycle.